### DISCUSSION TITLE
Additional selections necessary for ballots, or an implementation decision?

### DISCUSSION THREAD

**Discussion by JohnLCaron** [microsoft/electionguard#63] [Q&A] [answered]
For example, one of the standard input ballots in data/plaintext_ballots_simple.json:

```
    {
        "object_id": "25a7111b-4334-425a-87c1-f7a49f42b3a2",
        "ballot_style": "jefferson-county-ballot-style",
        "contests": [
            {
                "object_id": "justice-supreme-court",
                "ballot_selections": [
                    {
                        "object_id": "john-adams-selection",
                        "vote": 1
                    },
                    {
                        "object_id": "benjamin-franklin-selection",
                        "vote": 1
                    }
                ]
            }
        ]
    },
```
contains just the "voted for" selections.

If you look at the internals of the accepted ciphertext ballot, you see that 

1. For each possible selection on the contest, there is a selection in the accepted ballot (in this case 4).
2. In addition, there are L placeholder selections, where L = number of allowed selections for the contest (in this case 2).

If its a spoiled ballot, then when it gets decrypted into both a spoiled plaintext tally, and a spoiled plaintext ballot, all of those selections and placeholders are present.

**Questions:**

1) Is having all possible selections on an accepted ballot part of the spec, or is that just an implementation choice here? Since this is for every ballot, there would be significant memory and processing savings if only the voted for selections were needed.

2) The spec clearly states that one needs L placeholders added to each ballot. But for the sake of completeness, Ill ask the same question: would it be possible/desirable to change the spec to only add placeholders if the ballot is undervoted?

If the point of the spoiled plaintext tally and spoiled plaintext ballot is to allow observers to compare to the original ballot, Im wondering if seeing the placeholder selections might be confusing? So:

3) In the interest of clarity, could the placeholder selections be left out of the spoiled plaintext tally and spoiled plaintext ballot serializations?

4) In the interest of size, could the selections not voted for be left out of the spoiled plaintext tally and spoiled plaintext ballot serializations?

Ive decided to place this discussion in electionguard, rather than electionguard-python, because its an example of the question _What is the electionguard specification?_, versus _What is the electionguard python API?_

**Comment by rc-ms** [comment]
Great questions. Fundamentally, this convention is necessary for ballot well-formedness. The placeholder ballots are as you suggest an implementation choice to capture undervotes and in this case are _ **determined by the spec**_  I defer to @benaloh as to how he would prefer to treat this in the 1.0 spec (vs. mentioning other options), but this will still be the "default" implementation of the codebase.  I agree that upon spoilage that placeholder selections not be shown, as they would only confuse the voter.

**Comment by benaloh** [comment]
I went back and forth on the placeholders many times.  The issue is as follows:  one can either use the placeholders and have a simple NIZK that the total number of selections made on each contest is exactly L, or eliminate the placeholders and have a far more complex disjunctive NIZK that the total number of selections is an integer in the interval [0,L].  It turns out that the more complex NZIK requires almost exactly the same amount of space as placeholder values (because it is a disjunct of L+1 NIZKs).  So I eventually decided to use the placeholders because they are similar to what is already being done and doesn't introduce the new more complex construct into the code.  I still feel uncertain about this decision, but I think that it was probably right.

With this design, the placeholders must be included even if they are zeros (unselected).  Otherwise an observer could, simply by inspecting an encrypted ballot, determine how many selections a voter made in each contest.  While this may seem minor, even undervotes must be kept confidential.

As for what is displayed for spoiled ballots, there are two requirements:  the plaintext should match the voter expectations as closely as possible, and verifiers should be able to match the plaintext to the ciphertext as easily as possible.

**Comment by JohnLCaron** [comment]
Ok, sounds like the answer is to remove placeholder selections from spoiled tallies and ballots, and leave everything else as it is. 

I havent started thinking like a cryptographer, so I wasnt thinking about what observers could detect. Question 1 ("if only the voted-for selections were needed") would let them detect who a ballot voted for, probably not what we want. However, for the sake of considering future designs where efficiency was strongly needed, if the selection id (who they voted for)  and whether it was a placeholder were encrypted, then only and always storing exactly L selections might be possible. Not asking for an answer, just leaving breadcrumbs.

**Comment by benaloh** [comment]
Remember that at the time a ballot is encrypted and the tracking codes are produced, the device/code doing the encryption won't know whether a ballot will be cast or spoiled, so we can't just remove placeholders from spoiled ballots.  An election record needs to have all of the placeholders intact., but when we display the tallies and spoiled ballots in human-readable form, we can suppress the display of the placeholder values.  Verifiers need to check all of the placeholders, but they can -- when checking that the displayed text matches the record -- know that the displayed text should not include placeholders.

**Reply by JohnLCaron** [reply]
Yes, I understand. I should have said "remove placeholder selections from the published spoiled plaintext tallies and spoiled plaintext ballots".

**Comment by JohnLCaron** [comment]
I notice that in ballot_validator.ballot_is_valid_for_style(), an encrypted ballot is declared invalid if it doesnt contain a contest for every contest in the manifest for its style. So im just noting that "encrypted ballots must contain an entry for each contest in the manifest for the ballot's style, and an entry for each selection in that contest".

**Comment by JohnLCaron** [comment]
But I also see that each encrypted ballot contains more than the number of entries in the contest. If the number of valid entries is L, it contains L placeholders, in addition to one entry for each entry in the manifest.

Sorry if you've explained this, but I dont get it. Why are these extra placeholders needed, since there is already an entry for all selections, so an observer cant detect undervotes.

**Comment by JohnLCaron** [comment, marked as answer]
I found this comment, sorry for the noise.

class ContestDescriptionWithPlaceholders(ContestDescription):
    """
    ContestDescriptionWithPlaceholders is a `ContestDescription` with ElectionGuard `placeholder_selections`.
    (The ElectionGuard spec requires for n-of-m elections that there be *exactly* n counters that are one
    with the rest zero, so if a voter deliberately undervotes, one or more of the placeholder counters will
    become one. This allows the `ConstantChaumPedersenProof` to verify correctly for undervoted contests.)
    """

**Comment by JohnLCaron** [comment]
I think this one is clear, so Im going to close it.

**Comment by JohnLCaron** [comment]
I dont seem to have permission to close?

**Reply by rc-ms** [reply]
try now. just sent you invite?

**Reply by JohnLCaron** [reply]
Maybe I just dont know how. No close button, like for issues. "Lock conversation?"

**Comment by rc-ms** [comment]
Hmmm.. Yeah discussions be new and not in the standard manage access areas.  I’ll see what I can figure out

**Comment by JohnLCaron** [comment]
Trying "mark as answered".
